posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,324 @@
+
+## Comparison code
+
+```{r,warning=FALSE,message=FALSE}
+write.colors <- c(
+  "data.table::fwrite" = "#D6604D",
+  "pandas.to_csv " = "#BF812D"
+)
+file_path = 'data.csv'
+n.rows <- 100
+seconds.limit <- 10
+atime.write.vary.cols <- atime::atime(
+  N = as.integer(10^seq(2, 10, by = 0.2)),
+  setup = {
+    set.seed(1)
+    input.vec <- rnorm(n.rows * N)
+    input.mat <- matrix(input.vec, n.rows, N)
+    input.df <- data.frame(input.mat) 
+    pd <- import("pandas")
+    input_df_pd <- r_to_py(input.df)
+  },
+  seconds.limit = seconds.limit,
+  "data.table::fwrite" = {
+    data.table::fwrite(input.df, tempfile(), showProgress = FALSE)
+  },
+  "pandas.to_csv" = {
+    input_df_pd$to_csv(file_path, index = FALSE)
+  }
+)
+```
+
+```{r,warning=FALSE,message=FALSE}
+refs.write.vary.cols <- atime::references_best(atime.write.vary.cols)
+```
+
+The predict() function is used to generate predictions based on the atime::reference_best() dataset. The resulting plot illustrates the data size, N, that can be processed within a specific time or memory limit.
+
+```{r,warning=FALSE,message=FALSE}
+pred.write.vary.cols <- predict(refs.write.vary.cols)
+plot(pred.write.vary.cols)
+```
+
+```{r}
+gg.write.dt.pd <- plot(pred.write.vary.cols) +
+  theme(text = element_text(size = 15)) +
+  ggtitle(sprintf("Write real numbers to CSV, with pandas in Python \nand data.table in R, %d x N", n.rows)) +
+  scale_x_log10("N = number of columns to write") +
+  scale_y_log10("Computation time (seconds)\nmedian line, min/max band\nover 10 timings") +
+  facet_null() +
+  scale_fill_manual(values = write.colors) +
+  scale_color_manual(values = write.colors)
+```
+
+```
+print(gg.write.dt.pd)
+```
+
+The plot above shows that in terms of writing data, the data.table package in R outperforms the pandas library in Python and particularly useful when dealing with large datasets.
+
+# Example 2: Reading a CSV File with data.table::fread() and pandas.read_csv 
+
+## fread: fast CSV reader
+
+Data.table provides the fread() function for reading data from a CSV file while Pandas offers the read_csv() function for reading data from a CSV file.
+
+## Comparison code
+
+```{r,warning=FALSE,message=FALSE}
+read.colors <- c(
+  "data.table::fread" = "#D6604D",
+  "pandas.read_csv" = "#BF812D"
+)
+n.rows <- 100
+seconds.limit <- 10
+file_path = 'data.csv'
+atime.read <- atime::atime(
+  N = as.integer(10^seq(2, 15, by = 0.2)),
+  setup = {
+    set.seed(1)
+    input.vec <- rnorm(n.rows*N)
+    input.mat <- matrix(input.vec, n.rows, N)
+    input.df <- data.frame(input.mat)
+    input.csv <- tempfile()
+    fwrite(input.df, "data.csv")
+    pd <- import("pandas")
+    input_df_pd <- pd$DataFrame(input.df) 
+  },
+  seconds.limit = seconds.limit,
+  "data.table::fread" = {
+    data.table::fread("data.csv", showProgress = FALSE) 
+  },
+  "pandas.read_csv " = {
+    pd <- import("pandas")
+    reticulate::py_run_string("import pandas as pd")
```

Comment:
this should be moved to setup if possible

Reply:
alright